include/dice/template-library/switch_cases.hpp

```diff
@@ -1,82 +1,151 @@
 #ifndef HYPERTRIE_SWITCHTEMPLATEFUNCTIONS_HPP
 #define HYPERTRIE_SWITCHTEMPLATEFUNCTIONS_HPP
 
+#include <array>
 #include <algorithm>
 #include <cassert>
 #include <concepts>
 #include <cstdint>
+#include <utility>
 
 namespace dice::template_library {
 
 	namespace detail_switch_cases {
 
+		template<typename T1, typename T2>
+		struct common_integral;
+
+		template<std::integral T>
+		struct common_integral<T, T> {
+			using type = T;
+		};
+
+		template<std::unsigned_integral T1, std::unsigned_integral T2>
+		struct common_integral<T1, T2> {
+			using type = std::conditional_t<sizeof(T1) >= sizeof(T2), T1, T2>;
+		};
+
+		template<std::signed_integral T1, std::signed_integral T2>
+		struct common_integral<T1, T2> {
+			using type = std::conditional_t<sizeof(T1) >= sizeof(T2), T1, T2>;
+		};
+
+		template<std::signed_integral T1, std::unsigned_integral T2>
+		struct common_integral<T1, T2> {
+			using type = typename common_integral<T1, std::make_signed_t<T2>>::type;
+		};
+
+		template<std::unsigned_integral T1, std::signed_integral T2>
+		struct common_integral<T1, T2> {
+			using type = typename common_integral<std::make_signed_t<T1>, T2>::type;
+		};
+
+		template<typename Int, Int start, Int end, typename Acc>
+		struct make_integer_sequence;
+
+		template<typename Int, Int start, Int end, Int...acc>
+		struct make_integer_sequence<Int, start, end, std::integer_sequence<Int, acc...>> {
+			using type = typename make_integer_sequence<Int, start + 1, end, std::integer_sequence<Int, acc..., start>>::type;
+		};
+
+		template<typename Int, Int start, Int ...acc>
+		struct make_integer_sequence<Int, start, start, std::integer_sequence<Int, acc...>> {
+			using type = std::integer_sequence<Int, acc...>;
+		};
+
 		template<std::integral auto first, std::integral auto last>
 		struct Range {
-			using int_type = std::conditional_t < first < 0 or last<0, intmax_t, uintmax_t>;
+			using int_type = typename common_integral<decltype(first), decltype(last)>::type;
```

Comment:
Also notice this: the integral type being passed in is now the type you actually wanted instead of `(u)intmax_t`

Reply:
Only problem I see here is that it should not work for long and unsigned long because there is no signed C++-type with enough precision on 64-bit machines.